pkg/kubelet/cm/deviceplugin/device_plugin_handler.go

```diff
@@ -205,74 +406,73 @@ func (h *DevicePluginHandlerImpl) Allocate(pod *v1.Pod, container *v1.Container,
 		// requests may fail if we serve them in mixed order.
 		// TODO: may revisit this part later if we see inefficient resource allocation
 		// in real use as the result of this.
-		resp, err := h.devicePluginManager.Allocate(resource, append(devices.UnsortedList(), allocated...))
+		resp, err := h.devicePluginManager.Allocate(resource, allocDevices.UnsortedList())
 		if err != nil {
-			return nil, err
+			return err
+		}
+
+		// Update internal cached podDevices state.
+		h.Lock()
+		if oldDevices := h.podDevices.containerDevices(podUID, contName, resource); oldDevices != nil {
+			freeDevices := oldDevices.Difference(allocDevices)
+			glog.V(3).Infof("old %v alloc %v free devices %v for resource %s", oldDevices.UnsortedList(), allocDevices.UnsortedList(), freeDevices.UnsortedList(), resource)
+			h.allocatedDevices[resource] = h.allocatedDevices[resource].Difference(freeDevices)
 		}
-		ret = append(ret, resp)
+		h.podDevices.insert(podUID, contName, resource, allocDevices, resp)
+		h.Unlock()
 	}
+
 	// Checkpoints device to container allocation information.
-	if err := h.writeCheckpoint(); err != nil {
-		return nil, err
-	}
-	return ret, nil
+	return h.writeCheckpoint()
 }
 
-// updateAllocatedDevices updates the list of GPUs in use.
-// It gets a list of active pods and then frees any GPUs that are bound to
+// GetDeviceRunContainerOptions checks whether we have cached containerDevices
+// for the passed-in <pod, container> and returns its DeviceRunContainerOptions
+// for the found one. An empty struct is returned in case no cached state is found.
+func (h *HandlerImpl) GetDeviceRunContainerOptions(pod *v1.Pod, container *v1.Container) *DeviceRunContainerOptions {
+	h.Lock()
```

Comment:
I'm a bit surprised by the usage of the Mutex here. Usually you would want to keep critical section to a minimum.
We are more in a big lock situation here.

Reply:
I think this is needed to synchronize read/write access on h.podDevices. Also this part shouldn't be expensive.